File: r3e-deno/src/js/fhe.js
// Copyright @ 2023 - 2024, R3E Network
// All Rights Reserved

/**
 * @module fhe
 * @description Fully Homomorphic Encryption module for the R3E FaaS platform.
 */

import { core } from "./infra.js";
import { encode, decode } from "./encoding.js";

/**
 * Generate a key pair for FHE operations.
 * 
 * @param {string} schemeType - The type of FHE scheme to use (e.g., "TFHE", "OpenFHE").
 * @param {Object} parameters - Additional parameters for key generation.
 * @returns {string} The ID of the generated key pair.
 */
export function generateKeys(schemeType, parameters = {}) {
  return core.ops.op_fhe_generate_keys(schemeType, parameters);
}

/**
 * Encrypt data using a public key.
 * 
 * @param {string} publicKeyId - The ID of the public key.
 * @param {Uint8Array|string} plaintext - The data to encrypt (Uint8Array or string).
 * @returns {string} The ID of the generated ciphertext.
 */
export function encrypt(publicKeyId, plaintext) {
  // Convert string to Uint8Array if necessary
  const data = typeof plaintext === 'string' ? encode(plaintext) : plaintext;
  return core.ops.op_fhe_encrypt(publicKeyId, data);
}

/**
 * Decrypt data using a private key.
 * 
 * @param {string} privateKeyId - The ID of the private key.
 * @param {string} ciphertextId - The ID of the ciphertext.
 * @param {boolean} asString - Whether to return the result as a string (default: false).
 * @returns {Uint8Array|string} The decrypted data.
 */
export function decrypt(privateKeyId, ciphertextId, asString = false) {
  const data = core.ops.op_fhe_decrypt(privateKeyId, ciphertextId);
  return asString ? decode(data) : data;
}

/**
 * Add two ciphertexts homomorphically.
 * 
 * @param {string} ciphertext1Id - The ID of the first ciphertext.
 * @param {string} ciphertext2Id - The ID of the second ciphertext.
 * @returns {string} The ID of the resulting ciphertext.
 */
export function add(ciphertext1Id, ciphertext2Id) {
  return core.ops.op_fhe_add(ciphertext1Id, ciphertext2Id);
}

/**
 * Subtract one ciphertext from another homomorphically.
 * 
 * @param {string} ciphertext1Id - The ID of the first ciphertext.
 * @param {string} ciphertext2Id - The ID of the second ciphertext.
 * @returns {string} The ID of the resulting ciphertext.
 */
export function subtract(ciphertext1Id, ciphertext2Id) {
  return core.ops.op_fhe_subtract(ciphertext1Id, ciphertext2Id);
}

/**
 * Multiply two ciphertexts homomorphically.
 * 
 * @param {string} ciphertext1Id - The ID of the first ciphertext.
 * @param {string} ciphertext2Id - The ID of the second ciphertext.
 * @returns {string} The ID of the resulting ciphertext.
 */
export function multiply(ciphertext1Id, ciphertext2Id) {
  return core.ops.op_fhe_multiply(ciphertext1Id, ciphertext2Id);
}

/**
 * Negate a ciphertext homomorphically.
 * 
 * @param {string} ciphertextId - The ID of the ciphertext.
 * @returns {string} The ID of the resulting ciphertext.
 */
export function negate(ciphertextId) {
  return core.ops.op_fhe_negate(ciphertextId);
}

/**
 * Get a ciphertext by ID.
 * 
 * @param {string} ciphertextId - The ID of the ciphertext.
 * @returns {Object} The ciphertext object.
 */
export function getCiphertext(ciphertextId) {
  return core.ops.op_fhe_get_ciphertext(ciphertextId);
}

/**
 * Estimate the noise budget of a ciphertext.
 * 
 * @param {string} ciphertextId - The ID of the ciphertext.
 * @returns {number|null} The estimated noise budget, or null if not available.
 */
export function estimateNoiseBudget(ciphertextId) {
  return core.ops.op_fhe_estimate_noise_budget(ciphertextId);
}

/**
 * FHE scheme types supported by the platform.
 */
export const SchemeType = {
  TFHE: "TFHE",
  OPENFHE: "OpenFHE",
  SEAL: "SEAL",
  HELIB: "HElib",
  LATTIGO: "Lattigo",
};

/**
 * Example usage of the FHE module.
 * 
 * @example
 * ```javascript
 * import { fhe } from "r3e";
 * 
 * // Generate keys
 * const keyPairId = fhe.generateKeys(fhe.SchemeType.TFHE);
 * 
 * // Encrypt data
 * const plaintext = "Hello, FHE!";
 * const ciphertextId = fhe.encrypt(publicKeyId, plaintext);
 * 
 * // Perform homomorphic operations
 * const ciphertext2Id = fhe.encrypt(publicKeyId, "World");
 * const resultId = fhe.add(ciphertextId, ciphertext2Id);
 * 
 * // Decrypt the result
 * const result = fhe.decrypt(privateKeyId, resultId, true);
 * console.log(`Result: ${result}`);
 * 
 * // Check the noise budget
 * const noiseBudget = fhe.estimateNoiseBudget(resultId);
 * console.log(`Noise budget: ${noiseBudget}`);
 * ```
 */

File: r3e-deno/src/ext/fhe.rs
// Copyright @ 2023 - 2024, R3E Network
// All Rights Reserved

//! Fully Homomorphic Encryption operations for the R3E FaaS platform.

use crate::js_op;
use crate::sandbox::SandboxConfig;
use deno_core::error::AnyError;
use deno_core::op2;
use deno_core::OpState;
use r3e_fhe::{
    FheCiphertextId, FheError, FheKeyPairId, FheParameters, FhePrivateKeyId, FhePublicKeyId,
    FheResult, FheSchemeType, FheService, HomomorphicOperation,
};
use serde::{Deserialize, Serialize};
use std::sync::{Arc, Mutex};

/// Generate a key pair for FHE operations.
#[op2]
#[serde]
pub fn op_fhe_generate_keys(
    state: &mut OpState,
    #[serde] scheme_type: String,
    #[serde] parameters: serde_json::Value,
) -> Result<FheKeyPairId, AnyError> {
    // Check if the operation is allowed
    super::op_allowed("op_fhe_generate_keys", &serde_json::json!({
        "scheme_type": scheme_type,
    }))?;

    // Get the sandbox configuration
    let sandbox_config = state
        .borrow::<Arc<Mutex<SandboxConfig>>>()
        .lock()
        .map_err(|e| AnyError::msg(format!("Failed to acquire lock: {}", e)))?;

    // Check if the operation is allowed by the sandbox
    if !sandbox_config.allow_fhe_operations {
        return Err(AnyError::msg(
            "Fully Homomorphic Encryption operations are not allowed in this sandbox",
        ));
    }

    // Parse the scheme type
    let scheme_type = match scheme_type.as_str() {
        "TFHE" => FheSchemeType::Tfhe,
        "OpenFHE" => FheSchemeType::OpenFhe,
        "SEAL" => FheSchemeType::Seal,
        "HElib" => FheSchemeType::Helib,
        "Lattigo" => FheSchemeType::Lattigo,
        _ => {
            return Err(AnyError::msg(format!(
                "Unsupported FHE scheme type: {}",
                scheme_type
            )))
        }
    };

    // TODO: Implement actual FHE key generation
    // For now, we'll return a placeholder key pair ID
    let key_pair_id = FheKeyPairId::new();

    Ok(key_pair_id)
}

/// Encrypt data using a public key.
#[op2]
#[serde]
pub fn op_fhe_encrypt(
    state: &mut OpState,
    #[serde] public_key_id: FhePublicKeyId,
    #[serde] plaintext: Vec<u8>,
) -> Result<FheCiphertextId, AnyError> {
    // Check if the operation is allowed
    super::op_allowed("op_fhe_encrypt", &serde_json::json!({
        "public_key_id": public_key_id,
    }))?;

    // Get the sandbox configuration
    let sandbox_config = state
        .borrow::<Arc<Mutex<SandboxConfig>>>()
        .lock()
        .map_err(|e| AnyError::msg(format!("Failed to acquire lock: {}", e)))?;

    // Check if the operation is allowed by the sandbox
    if !sandbox_config.allow_fhe_operations {
        return Err(AnyError::msg(
            "Fully Homomorphic Encryption operations are not allowed in this sandbox",
        ));
    }

    // TODO: Implement actual FHE encryption
    // For now, we'll return a placeholder ciphertext ID
    let ciphertext_id = FheCiphertextId::new();

    Ok(ciphertext_id)
}

/// Decrypt data using a private key.
#[op2]
#[serde]
pub fn op_fhe_decrypt(
    state: &mut OpState,
    #[serde] private_key_id: FhePrivateKeyId,
    #[serde] ciphertext_id: FheCiphertextId,
) -> Result<Vec<u8>, AnyError> {
    // Check if the operation is allowed
    super::op_allowed("op_fhe_decrypt", &serde_json::json!({
        "private_key_id": private_key_id,
        "ciphertext_id": ciphertext_id,
    }))?;

    // Get the sandbox configuration
    let sandbox_config = state
        .borrow::<Arc<Mutex<SandboxConfig>>>()
        .lock()
        .map_err(|e| AnyError::msg(format!("Failed to acquire lock: {}", e)))?;

    // Check if the operation is allowed by the sandbox
    if !sandbox_config.allow_fhe_operations {
        return Err(AnyError::msg(
            "Fully Homomorphic Encryption operations are not allowed in this sandbox",
        ));
    }

    // TODO: Implement actual FHE decryption
    // For now, we'll return a placeholder plaintext
    let plaintext = vec![0, 1, 2, 3];

    Ok(plaintext)
}

/// Add two ciphertexts homomorphically.
#[op2]
#[serde]
pub fn op_fhe_add(
    state: &mut OpState,
    #[serde] ciphertext1_id: FheCiphertextId,
    #[serde] ciphertext2_id: FheCiphertextId,
) -> Result<FheCiphertextId, AnyError> {
    // Check if the operation is allowed
    super::op_allowed("op_fhe_add", &serde_json::json!({
        "ciphertext1_id": ciphertext1_id,
        "ciphertext2_id": ciphertext2_id,
    }))?;

    // Get the sandbox configuration
    let sandbox_config = state
        .borrow::<Arc<Mutex<SandboxConfig>>>()
        .lock()
        .map_err(|e| AnyError::msg(format!("Failed to acquire lock: {}", e)))?;

    // Check if the operation is allowed by the sandbox
    if !sandbox_config.allow_fhe_operations {
        return Err(AnyError::msg(
            "Fully Homomorphic Encryption operations are not allowed in this sandbox",
        ));
    }

    // TODO: Implement actual FHE addition
    // For now, we'll return a placeholder ciphertext ID
    let result_id = FheCiphertextId::new();

    Ok(result_id)
}

/// Subtract one ciphertext from another homomorphically.
#[op2]
#[serde]
pub fn op_fhe_subtract(
    state: &mut OpState,
    #[serde] ciphertext1_id: FheCiphertextId,
    #[serde] ciphertext2_id: FheCiphertextId,
) -> Result<FheCiphertextId, AnyError> {
    // Check if the operation is allowed
    super::op_allowed("op_fhe_subtract", &serde_json::json!({
        "ciphertext1_id": ciphertext1_id,
        "ciphertext2_id": ciphertext2_id,
    }))?;

    // Get the sandbox configuration
    let sandbox_config = state
        .borrow::<Arc<Mutex<SandboxConfig>>>()
        .lock()
        .map_err(|e| AnyError::msg(format!("Failed to acquire lock: {}", e)))?;

    // Check if the operation is allowed by the sandbox
    if !sandbox_config.allow_fhe_operations {
        return Err(AnyError::msg(
            "Fully Homomorphic Encryption operations are not allowed in this sandbox",
        ));
    }

    // TODO: Implement actual FHE subtraction
    // For now, we'll return a placeholder ciphertext ID
    let result_id = FheCiphertextId::new();

    Ok(result_id)
}

/// Multiply two ciphertexts homomorphically.
#[op2]
#[serde]
pub fn op_fhe_multiply(
    state: &mut OpState,
    #[serde] ciphertext1_id: FheCiphertextId,
    #[serde] ciphertext2_id: FheCiphertextId,
) -> Result<FheCiphertextId, AnyError> {
    // Check if the operation is allowed
    super::op_allowed("op_fhe_multiply", &serde_json::json!({
        "ciphertext1_id": ciphertext1_id,
        "ciphertext2_id": ciphertext2_id,
    }))?;

    // Get the sandbox configuration
    let sandbox_config = state
        .borrow::<Arc<Mutex<SandboxConfig>>>()
        .lock()
        .map_err(|e| AnyError::msg(format!("Failed to acquire lock: {}", e)))?;

    // Check if the operation is allowed by the sandbox
    if !sandbox_config.allow_fhe_operations {
        return Err(AnyError::msg(
            "Fully Homomorphic Encryption operations are not allowed in this sandbox",
        ));
    }

    // TODO: Implement actual FHE multiplication
    // For now, we'll return a placeholder ciphertext ID
    let result_id = FheCiphertextId::new();

    Ok(result_id)
}

/// Negate a ciphertext homomorphically.
#[op2]
#[serde]
pub fn op_fhe_negate(
    state: &mut OpState,
    #[serde] ciphertext_id: FheCiphertextId,
) -> Result<FheCiphertextId, AnyError> {
    // Check if the operation is allowed
    super::op_allowed("op_fhe_negate", &serde_json::json!({
        "ciphertext_id": ciphertext_id,
    }))?;

    // Get the sandbox configuration
    let sandbox_config = state
        .borrow::<Arc<Mutex<SandboxConfig>>>()
        .lock()
        .map_err(|e| AnyError::msg(format!("Failed to acquire lock: {}", e)))?;

    // Check if the operation is allowed by the sandbox
    if !sandbox_config.allow_fhe_operations {
        return Err(AnyError::msg(
            "Fully Homomorphic Encryption operations are not allowed in this sandbox",
        ));
    }

    // TODO: Implement actual FHE negation
    // For now, we'll return a placeholder ciphertext ID
    let result_id = FheCiphertextId::new();

    Ok(result_id)
}

/// Get a ciphertext by ID.
#[op2]
#[serde]
pub fn op_fhe_get_ciphertext(
    state: &mut OpState,
    #[serde] ciphertext_id: FheCiphertextId,
) -> Result<serde_json::Value, AnyError> {
    // Check if the operation is allowed
    super::op_allowed("op_fhe_get_ciphertext", &serde_json::json!({
        "ciphertext_id": ciphertext_id,
    }))?;

    // Get the sandbox configuration
    let sandbox_config = state
        .borrow::<Arc<Mutex<SandboxConfig>>>()
        .lock()
        .map_err(|e| AnyError::msg(format!("Failed to acquire lock: {}", e)))?;

    // Check if the operation is allowed by the sandbox
    if !sandbox_config.allow_fhe_operations {
        return Err(AnyError::msg(
            "Fully Homomorphic Encryption operations are not allowed in this sandbox",
        ));
    }

    // TODO: Implement actual FHE ciphertext retrieval
    // For now, we'll return a placeholder ciphertext
    let ciphertext = serde_json::json!({
        "id": ciphertext_id,
        "scheme_type": "TFHE",
        "public_key_id": FhePublicKeyId::new(),
        "created_at": 1614556800,
        "metadata": {
            "plaintext_size": 4,
            "ciphertext_size": 1024,
            "operation_count": 0,
            "noise_budget": 120,
            "properties": {}
        }
    });

    Ok(ciphertext)
}

/// Estimate the noise budget of a ciphertext.
#[op2]
#[serde]
pub fn op_fhe_estimate_noise_budget(
    state: &mut OpState,
    #[serde] ciphertext_id: FheCiphertextId,
) -> Result<Option<u32>, AnyError> {
    // Check if the operation is allowed
    super::op_allowed("op_fhe_estimate_noise_budget", &serde_json::json!({
        "ciphertext_id": ciphertext_id,
    }))?;

    // Get the sandbox configuration
    let sandbox_config = state
        .borrow::<Arc<Mutex<SandboxConfig>>>()
        .lock()
        .map_err(|e| AnyError::msg(format!("Failed to acquire lock: {}", e)))?;

    // Check if the operation is allowed by the sandbox
    if !sandbox_config.allow_fhe_operations {
        return Err(AnyError::msg(
            "Fully Homomorphic Encryption operations are not allowed in this sandbox",
        ));
    }

    // TODO: Implement actual FHE noise budget estimation
    // For now, we'll return a placeholder noise budget
    Ok(Some(120))
}
